src/txn/txn_timestamp.c

```diff
@@ -713,25 +713,51 @@ __wt_txn_set_prepare_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t prepare_
     WT_RET(__txn_assert_after_reads(session, "prepare", prepare_ts));
 
     /*
-     * Check whether the prepare timestamp is less than the oldest timestamp.
+     * Check whether the prepare timestamp is less than the stable timestamp.
      */
-    oldest_ts = txn_global->oldest_timestamp;
-    if (prepare_ts < oldest_ts) {
+    stable_ts = txn_global->stable_timestamp;
+    if (prepare_ts <= stable_ts) {
         /*
-         * Check whether the prepare timestamp needs to be rounded up to the oldest timestamp.
+         * Check whether the application is using the "prepared" roundup mode. This rounds up to
+         * _oldest_, not stable, and permits preparing before stable, because it is meant to be used
+         * during application recovery to replay a transaction that was successfully prepared (and
+         * possibly committed) before a crash but had not yet become durable. In general it is
+         * important to replay such transactions at the same time they had before the crash; in a
+         * distributed setting they might have already committed in the network, in which case
+         * replaying them at a different time is very likely to be inconsistent. Meanwhile, once a
+         * transaction prepares we allow stable to move forward past it, so replaying may require
+         * preparing and even committing prior to stable.
+         *
+         * Such a replay is safe provided that it happens during application-level recovery before
+         * resuming ordinary operations: between the time the transaction prepares and the crash,
+         * operations intersecting with the prepared transaction fail with WT_PREPARE_CONFLICT, and
+         * after the crash, the replay recreates this state before any ordinary operations can
+         * intersect with it. Application recovery code is responsible for making sure that any
+         * other operations it does before the replay that might intersect with the prepared
+         * transaction are consistent with it.
+         *
+         * (There is a slight extra wrinkle at the moment, because it is possible for a transaction
+         * to prepare and commit and be interacted with before it becomes durable. Currently such
+         * transactions _must_ be replayed identically by the application to avoid inconsistency,
+         * or avoided. FIXME-WT-8747: remove this note when WT-8747 fixes this.)
+         *
+         * In general preparing or committing before stable is unsafe and can cause data consistency
+         * failures, so use of "roundup=prepared" for other purposes is strongly discouraged.
```

Comment:
I don't think the judgment here is helpful. I think calling this a recovery-only feature is sufficient. Because prepared updates are not durable, this API is unfortunately necessary. Saying it's discouraged sounds like there's an encouraged alternative. I'd instead state:

```
This flag is only intended to be used for application level recovery when ordinary operations are locked out.
```

Which is basically paragraph 2. So maybe touching up its introduction:
```
Such a replay is safe provided that it happens during application-level recovery before resuming ordinary operations
```
to instead say:
```
Such a replay is intended for application-level recovery and, for correctness, must be done before resuming ordinary operations
```

Reply:
We/I want to be explicit about this being unsafe except in the prescribed circumstances. I rearranged the last paragraph to be less judgy; let me know if it's better.

(and, sorry this note got delayed, I wrote something here and apparently forgot to post it)